src/vtt-segment-loader.js

```diff
@@ -1006,44 +972,166 @@ export default class SegmentLoader extends videojs.EventTarget {
+      this.pendingSegment_ = null;
+      this.state = 'READY';
+      return;
     }
 
     segmentInfo.byteLength = segmentInfo.bytes.byteLength;
+
     if (typeof segment.start === 'number' && typeof segment.end === 'number') {
       this.mediaSecondsLoaded += segment.end - segment.start;
     } else {
       this.mediaSecondsLoaded += segment.duration;
     }
 
-    this.sourceUpdater_.appendBuffer(segmentInfo.bytes,
-                                     this.handleUpdateEnd_.bind(this));
+    segmentInfo.cues.forEach((cue) => {
+      this.subtitlesTrack_.addCue(cue);
+    });
+
+    this.handleUpdateEnd_();
+  }
+
+  parseVTTCues_(segmentInfo) {
+    let decoder;
+    let decodeBytesToString = false;
+
+    if (typeof window.TextDecoder === 'function') {
+      decoder = new window.TextDecoder('utf8');
+    } else {
+      decoder = window.WebVTT.StringDecoder();
+      decodeBytesToString = true;
+    }
+
+    const parser = new window.WebVTT.Parser(window,
+                                            window.vttjs,
+                                            decoder);
+
+    segmentInfo.cues = [];
+    segmentInfo.timestampmap = { MPEGTS: 0, LOCAL: 0 };
+
+    parser.oncue = segmentInfo.cues.push.bind(segmentInfo.cues);
+    parser.ontimestampmap = (map) => segmentInfo.timestampmap = map;
+    parser.onparsingerror = (error) => {
+      videojs.log.warn('Error encountered when parsing cues: ' + error.message);
+    };
+
+    if (segmentInfo.segment.map) {
+      let mapData = segmentInfo.segment.map.bytes;
+
+      if (decodeBytesToString) {
+        mapData = uintToString(mapData);
+      }
+
+      parser.parse(mapData);
+    }
+
+    let segmentData = segmentInfo.bytes;
+
+    if (decodeBytesToString) {
+      segmentData = uintToString(segmentData);
+    }
+
+    parser.parse(segmentData);
+    parser.flush();
+  }
+
+  updateTimeMapping_(segmentInfo, mappingObj, playlist) {
+    let segment = segmentInfo.segment;
+    let timestampmap = segmentInfo.timestampmap;
+
+    if (!mappingObj || !segmentInfo.cues.length) {
+      // If the sync controller does not have a mapping of TS to Media Time for the
+      // timeline, then we don't have enough information to update the segment and cue
+      // start/end times
+      // If there are no cues, we also do not have enough information to figure out
+      // segment timing
+      return;
+    }
+
+    const diff = (timestampmap.MPEGTS / 90000) - timestampmap.LOCAL + mappingObj.mapping;
+
+    segmentInfo.cues.forEach((cue) => {
+      // First convert cue time to TS time using the timestamp-map provided within the vtt
+      cue.startTime += diff;
+      cue.endTime += diff;
+    });
+
+    const firstStart = segmentInfo.cues[0].startTime;
+    const lastStart = segmentInfo.cues[segmentInfo.cues.length - 1].startTime;
+    const midPoint = (firstStart + lastStart) / 2;
+
+    segment.start = midPoint - (segment.duration / 2);
+    segment.end = midPoint + (segment.duration / 2);
```

Comment:
We decided to try not storing segment start and end times and letting the conservative guess make it work since web vtt segments should be much smaller than video or audio segments.

One thing to be careful about is that the "resync" functionality needs to be removed since it will continuously reset the segment loader on poor guess instead of just walking forward.